### PR TITLE
backends: Add SDL2 Support

### PR DESCRIPTION
**Pull request by RobLoach**
This adds an SDL2 backend for GFXprim. I'm not sure if it's the best approach yet, but it is building and running in SDL2!

### CODE REVIEW

**Comment by RobLoach** [comment]
Need to add sdl2 deps to GitHub Actions, and update a test.

**Comment by metan-ucw** [comment]
I've quickly looked at the code and it generally looks good.

There are two minor things to fix so far:

* There seems to be unrelated Makefile changes mixed in, i.e. the code that disables widgets on missing json-c
* Newer code should generally use `if (!foo)` instead of `if (foo == NULL)`

**Comment by metan-ucw** [comment]
> Need to add sdl2 deps to GitHub Actions, and update a test.

You can add SDL2 to the actions as a part of this patchset, there is a `.github/workflow/tests.yml` file and all that needs to be done is to add sdl2 dev ubuntu package to the build-with-libs run.

File: doc/backends.txt [hunk 1]
```diff
@@ -79,6 +79,32 @@ The caption is window caption.
...
 window resizable.
 
+SDL2
+~~~~
+
+[source,c]
+-------------------------------------------------------------------------------
+enum gp_sdl2_flags {
+        GP_SDL_FULLSCREEN = 0x01,
+        GP_SDL_RESIZABLE  = 0x02,
```
Comment (RobLoach) [inline doc/backends.txt:89]:
``` suggestion
        GP_SDL2_FULLSCREEN = 0x01,
        GP_SDL2_RESIZABLE  = 0x02,
```

**Comment by metan-ucw** [comment]
Code looks good, there are some minor misplaced whitespaces but that can be fixed before applying.

However looking at SDL and SDL2 the the API looks quite similar so I guess that we can link only against one. My guess is that we will get linker failures if we attempt to link backends library against both of them. Hence we may need to tweak the configure script to check for SDL only if SDL2 is not available.

**Comment by RobLoach** [comment]
> However looking at SDL and SDL2 the the API looks quite similar

Yes, was originally considering having the two merged, and check which version is active to flip between the two APIs. Ended up having them as separate implementations because there are differences in linking, and subtle changes in the API.

**Comment by metan-ucw** [comment]
@RobLoach Hmm, as long as we make it so that only one of them is enabled it may be doable to add SDL2 support to the current backend with a few ifdefs. And I guess that the gp_pixmap_from_sdl_surface() is the only thing than has to have explicitly versioned and ther has to be sdl sdl2 version and at least one of them would return ENOSYS at any time.

**Comment by metan-ucw** [comment]
I've tried the approach with ifdefs and pushed work-in-progress branch at https://github.com/gfxprim/gfxprim/tree/SDL2-support

It's probably not even working but it should show the direction.

**Comment by RobLoach** [comment]
Nice work, getting an error:
```
➜  c_simple git:(SDL2-support) LD_LIBRARY_PATH=../../build ./SDL_glue               

C stack trace (most recent call first):
gfxprim/build/libgfxprim.so.1(gp_debug_print_cstack+0x32)[0x7f76af48e092]
gfxprim/build/libgfxprim.so.1(gp_debug_print+0x37d)[0x7f76af48f30d]
gfxprim/build/libgfxprim-backends.so.1(gp_pixmap_from_sdl_surface+0x2e)[0x7f76af39117e]
./SDL_glue(main+0x49)[0x55b96a75a269]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7f76af1be0b3]
./SDL_glue(+0x134e)[0x55b96a75a34e]
*** FATAL: gp_sdl.c:gp_pixmap_from_sdl_surface():147: SDL support not compiled in.
*** gfxprim: gp_convert.gen.c:23: in gp_RGB888_to_pixel: 

Cannot convert RGB888 to GP_PIXEL_UNKNOWN

C stack trace (most recent call first):
gfxprim/build/libgfxprim.so.1(gp_debug_print_cstack+0x32)[0x7f76af48e092]
gfxprim/build/libgfxprim.so.1(gp_print_abort_info+0x115)[0x7f76af48e215]
gfxprim/build/libgfxprim.so.1(gp_RGB888_to_pixel+0x2a2)[0x7f76af48e512]
./SDL_glue(main+0x56)[0x55b96a75a276]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7f76af1be0b3]
./SDL_glue(+0x134e)[0x55b96a75a34e]
[1]    4594 abort (core dumped)  LD_LIBRARY_PATH=../../build ./SDL_glue
```

For reference, I have both SDL1 and SDL2 libraries available on this machine. `LIBSDL_VERSION` was set to 2.

**Comment by metan-ucw** [comment]
This specific failure is caused by the fact that there is no error checking from gp_pixmap_from_sdl_surface() and we continue regardless...

As I've said this is work-in-progress and probably broken even more than this. I will try to finish the patch once I have a bit of time...

**Comment by RobLoach** [comment]
Still tweaking this. I've split it further apart to see what it would potentially look like. There are still things I believe we can share between the two, while still having the ability to have both enabled, or either.

Not sure how to state this as a draft.

**Comment by metan-ucw** [comment]
I do not think that we can have both SDL 1.2 and SDL 2 linked in the gfxprim backends at the same time. The problem is that they have functions with exactly same names, this will yield a linker failure when libgfxprim-backends.so is being build. I guess that the only option would be playing tricks with dlopen() which is something I would like to avoid if possible.

**Comment by RobLoach** [comment]
>  I guess that the only option would be playing tricks with dlopen() which is something I would like to avoid if possible.

Agreed about avoiding `dlopen()`. Another option may be [sdl12-compat](https://github.com/libsdl-org/sdl12-compat), but I'm not sure how viable that is.

> I do not think that we can have both SDL 1.2 and SDL 2 linked in the gfxprim backends at the same time.

Both SDL_Glue and SDL2_Glue seem to be working here, while gfxprim was compiled with both.

sdl1 sdl2

**Comment by metan-ucw** [comment]
The SDL_glue is a bit different, that program does not link against backends at all. I've just realized that we can put the function gp_pixmap_from_sdl_sufrace() into a header as static inline and it will work regardless of the SDL version.

For backends it's different case though, since the libgfxprim-backends.so has to link against the libraries it uses and as far as I can tell linking against both SDL1.2 and SDL2 is not possible.

**Comment by RobLoach** [comment]
Nice work on that new SDL2 support branch. I think combining them is the right approach here :+1:

**Comment by metan-ucw** [comment]
FYI I've pushed SDL2 support that seems to work fine. Please test and report bugs if there are any.

**Pull request closed without merging**